File: scripts/dev_tools/pr_context/verification_evidence.py
"""Discover and parse canonical feature verification evidence artifacts.

Purpose:
    Provide deterministic discovery and strict schema parsing for canonical
    evidence markdown files so PR context generation can make traceable
    verification claims.

Flow:
    1. Discover canonical evidence files under active feature folders.
    2. Parse required schema fields from each markdown file.
    3. Normalize pass/fail/unparseable status from EXIT_CODE.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING, Literal

if TYPE_CHECKING:
    from pathlib import Path

REQUIRED_FIELDS: tuple[str, str, str] = ("Timestamp", "Command", "EXIT_CODE")
EXPECTED_EXIT_CODE_FIELD: str = "ExpectedExitCode"
CANONICAL_GLOBS: tuple[str, str, str] = (
    "evidence/qa-gates/**/*.md",
    "evidence/regression-testing/**/*.md",
    "evidence/other/**/*.md",
)

NormalizedResult = Literal["pass", "fail", "unparseable"]


@dataclass(frozen=True)
class VerificationEvidenceRecord:
    """Represent one parsed canonical evidence artifact.

    Purpose:
        Carry all fields needed to render verification rows and preserve source
        traceability to a specific feature evidence file.

    Attributes:
        feature: Active feature folder identifier.
        source_file: Repository-relative evidence file path.
        timestamp: Parsed `Timestamp` field when present.
        command: Parsed `Command` field when present.
        exit_code: Parsed `EXIT_CODE` integer when parseable.
        normalized_result: Deterministic status (`pass`, `fail`, `unparseable`).
        expected_exit_code: Declared expected exit code, `0` when the optional
            `ExpectedExitCode` field is absent and on every `unparseable` path.
    """

    feature: str
    source_file: str
    timestamp: str | None
    command: str | None
    exit_code: int | None
    normalized_result: NormalizedResult
    expected_exit_code: int = 0


def normalize_result(exit_code: int, expected_exit_code: int) -> NormalizedResult:
    """Normalize an observed exit code against its declared expectation.

    Args:
        exit_code: Observed process exit code parsed from `EXIT_CODE`.
        expected_exit_code: Declared expectation, `0` when undeclared.

    Returns:
        `pass` when the observed code equals the expectation, `fail` otherwise.

    Side Effects:
        None.
    """
    return "pass" if exit_code == expected_exit_code else "fail"


def discover_canonical_evidence_files(root: Path, feature: str) -> list[Path]:
    """Discover canonical evidence files for one active feature.

    Args:
        root: Repository root path.
        feature: Active feature directory name under `docs/features/active`.

    Returns:
        Sorted, deduplicated repository-relative paths for canonical evidence files.

    Side Effects:
        Reads filesystem metadata through globbing.
    """
    feature_root = root / "docs" / "features" / "active" / feature
    if not feature_root.exists() or not feature_root.is_dir():
        return []

    discovered: set[Path] = set()
    # Search canonical evidence roots in a fixed order, then sort for stability.
    for pattern in CANONICAL_GLOBS:
        for candidate in feature_root.glob(pattern):
            if candidate.is_file():
                discovered.add(candidate.relative_to(root))
    return sorted(discovered)


def parse_verification_evidence_markdown(
    *, feature: str, source_file: str, markdown: str
) -> VerificationEvidenceRecord:
    """Parse required schema fields and normalize verification status.

    Args:
        feature: Active feature identifier owning the evidence file.
        source_file: Repository-relative evidence file path.
        markdown: Raw markdown content to parse.

    Returns:
        A normalized evidence record with `pass`, `fail`, or `unparseable` result.

    Side Effects:
        None.
    """
    parsed: dict[str, str] = {}

    # Parse `Key: value` rows once and keep only required schema fields.
    for raw_line in markdown.splitlines():
        if ":" not in raw_line:
            continue
        key, value = raw_line.split(":", 1)
        key = key.strip()
        if key in REQUIRED_FIELDS:
            parsed[key] = value.strip()
        elif key == EXPECTED_EXIT_CODE_FIELD and key not in parsed:
            parsed[key] = value.strip()

    timestamp = parsed.get("Timestamp")
    command = parsed.get("Command")
    exit_code_raw = parsed.get("EXIT_CODE")
    expected_exit_code_raw = parsed.get(EXPECTED_EXIT_CODE_FIELD)

    if not timestamp or not command or exit_code_raw is None:
        return VerificationEvidenceRecord(
            feature=feature,
            source_file=source_file,
            timestamp=timestamp,
            command=command,
            exit_code=None,
            normalized_result="unparseable",
            expected_exit_code=0,
        )

    try:
        exit_code = int(exit_code_raw)
    except ValueError:
        return VerificationEvidenceRecord(
            feature=feature,
            source_file=source_file,
            timestamp=timestamp,
            command=command,
            exit_code=None,
            normalized_result="unparseable",
            expected_exit_code=0,
        )

    # An undeclared expectation defaults to zero, reproducing prior behavior.
    try:
        expected_exit_code = (
            0 if expected_exit_code_raw is None else int(expected_exit_code_raw)
        )
    except ValueError:
        return VerificationEvidenceRecord(
            feature=feature,
            source_file=source_file,
            timestamp=timestamp,
            command=command,
            exit_code=None,
            normalized_result="unparseable",
            expected_exit_code=0,
        )

    normalized_result: NormalizedResult = normalize_result(
        exit_code, expected_exit_code
    )
    return VerificationEvidenceRecord(
        feature=feature,
        source_file=source_file,
        timestamp=timestamp,
        command=command,
        exit_code=exit_code,
        normalized_result=normalized_result,
        expected_exit_code=expected_exit_code,
    )


def parse_verification_evidence_file(
    *, root: Path, feature: str, relative_path: Path
) -> VerificationEvidenceRecord:
    """Read and parse one canonical evidence file.

    Args:
        root: Repository root path.
        feature: Active feature identifier.
        relative_path: Repository-relative evidence file path.

    Returns:
        Parsed normalized evidence record.

    Raises:
        OSError: Propagated when the evidence file cannot be read.

    Side Effects:
        Reads evidence file content from disk.
    """
    markdown = (root / relative_path).read_text(encoding="utf-8")
    return parse_verification_evidence_markdown(
        feature=feature,
        source_file=relative_path.as_posix(),
        markdown=markdown,
    )

File: extensions/drm-copilot/src/lib/pr-context/verification-evidence.ts
/**
 * Discover and parse canonical feature verification evidence artifacts.
 *
 * Purpose:
 *     Port of `dev_tools/pr_context/verification_evidence.py`. Provide
 *     deterministic discovery and strict schema parsing for canonical evidence
 *     markdown files so PR context generation can make traceable verification
 *     claims.
 *
 * Flow:
 *     1. Discover canonical evidence files under an active feature folder.
 *     2. Parse required schema fields from each markdown file.
 *     3. Normalize pass/fail/unparseable status from `EXIT_CODE`.
 *
 * Responsibilities:
 *     - All filesystem access flows through the injected {@link FileSystem}
 *       (`exists`/`isDirectory`/`glob`/`readTextFile`) so the module is
 *       hermetic.
 */

import { type FileSystem, toPosixPath } from "../file-system";

/** Required schema fields parsed from an evidence markdown file. */
export const REQUIRED_FIELDS = ["Timestamp", "Command", "EXIT_CODE"] as const;

/** Optional schema field declaring the exit code a gate is expected to produce. */
export const EXPECTED_EXIT_CODE_FIELD = "ExpectedExitCode";

/** Canonical evidence glob roots, searched in fixed order. */
export const CANONICAL_GLOBS = [
  "evidence/qa-gates/**/*.md",
  "evidence/regression-testing/**/*.md",
  "evidence/other/**/*.md",
] as const;

/** Deterministic verification result derived from `EXIT_CODE`. */
export type NormalizedResult = "pass" | "fail" | "unparseable";

/** One parsed canonical evidence artifact. */
export interface VerificationEvidenceRecord {
  readonly feature: string;
  readonly sourceFile: string;
  readonly timestamp: string | null;
  readonly command: string | null;
  readonly exitCode: number | null;
  readonly normalizedResult: NormalizedResult;
  /**
   * Declared expected exit code. Defaults to `0` when the optional
   * `ExpectedExitCode` field is absent, and is `0` on every `unparseable` path.
   */
  readonly expectedExitCode: number;
}

/**
 * Normalize an observed exit code against its declared expectation.
 *
 * Mirrors Python `normalize_result`. Pure: performs no I/O.
 *
 * @param exitCode Observed process exit code parsed from `EXIT_CODE`.
 * @param expectedExitCode Declared expectation, `0` when undeclared.
 * @returns `pass` when the observed code equals the expectation, `fail` otherwise.
 */
export function normalizeResult(
  exitCode: number,
  expectedExitCode: number,
): NormalizedResult {
  return exitCode === expectedExitCode ? "pass" : "fail";
}

/**
 * Discover canonical evidence files for one active feature.
 *
 * Mirrors Python `discover_canonical_evidence_files`: when the feature root is
 * absent or not a directory, returns `[]`; otherwise globs the three canonical
 * roots (fixed order), keeps files, deduplicates, and returns repo-relative
 * POSIX paths sorted by code point.
 *
 * @param fs Injected filesystem.
 * @param root Repository root path.
 * @param feature Active feature directory name under `docs/features/active`.
 * @returns Sorted, deduplicated repo-relative POSIX evidence paths.
 */
export function discoverCanonicalEvidenceFiles(
  fs: FileSystem,
  root: string,
  feature: string,
): string[] {
  const normalizedRoot = toPosixPath(root).replace(/\/+$/u, "");
  const featureRoot = `${normalizedRoot}/docs/features/active/${feature}`;
  if (!fs.exists(featureRoot) || !fs.isDirectory(featureRoot)) {
    return [];
  }

  const discovered = new Set<string>();
  // Search canonical evidence roots in a fixed order, then sort for stability.
  for (const pattern of CANONICAL_GLOBS) {
    for (const candidate of fs.glob(featureRoot, pattern)) {
      if (fs.isFile(candidate)) {
        discovered.add(relativeToPosix(normalizedRoot, candidate));
      }
    }
  }
  return [...discovered].sort(compareCodePoint);
}

/**
 * Parse required schema fields and normalize verification status.
 *
 * Mirrors Python `parse_verification_evidence_markdown`: parse `Key: value`
 * rows (first occurrence wins for required fields), and when any required field
 * is missing or `EXIT_CODE` is not an integer, the result is `unparseable`;
 * otherwise `pass` when exit code is `0`, else `fail`.
 *
 * @param params Feature id, source path, and raw markdown content.
 * @returns A normalized evidence record.
 */
export function parseVerificationEvidenceMarkdown(params: {
  feature: string;
  sourceFile: string;
  markdown: string;
}): VerificationEvidenceRecord {
  const { feature, sourceFile, markdown } = params;
  const parsed = new Map<string, string>();

  // Parse `Key: value` rows once, keeping only the first occurrence of each
  // required schema field (mirrors the Python dict-first-write semantics).
  for (const rawLine of splitLines(markdown)) {
    const colonIndex = rawLine.indexOf(":");
    if (colonIndex === -1) {
      continue;
    }
    const key = rawLine.slice(0, colonIndex).trim();
    const value = rawLine.slice(colonIndex + 1).trim();
    if (
      (REQUIRED_FIELDS as readonly string[]).includes(key) &&
      !parsed.has(key)
    ) {
      parsed.set(key, value);
    }
    // Separate `if` rather than `else if` so the required-field block above is
    // left byte-identical; the two conditions are mutually exclusive because the
    // optional expectation field is not one of the required fields.
    if (key === EXPECTED_EXIT_CODE_FIELD && !parsed.has(key)) {
      parsed.set(key, value);
    }
  }

  const timestamp = parsed.get("Timestamp") ?? null;
  const command = parsed.get("Command") ?? null;
  const exitCodeRaw = parsed.get("EXIT_CODE");
  const expectedExitCodeRaw = parsed.get(EXPECTED_EXIT_CODE_FIELD);

  // A missing required field cannot produce a verifiable result.
  if (!timestamp || !command || exitCodeRaw === undefined) {
    return {
      feature,
      sourceFile,
      timestamp,
      command,
      exitCode: null,
      normalizedResult: "unparseable",
      expectedExitCode: 0,
    };
  }

  const exitCode = parseIntegerStrict(exitCodeRaw);
  if (exitCode === null) {
    return {
      feature,
      sourceFile,
      timestamp,
      command,
      exitCode: null,
      normalizedResult: "unparseable",
      expectedExitCode: 0,
    };
  }

  // An undeclared expectation defaults to zero, reproducing prior behavior; a
  // declared but non-integer expectation is as unparseable as a bad EXIT_CODE.
  const expectedExitCode =
    expectedExitCodeRaw === undefined
      ? 0
      : parseIntegerStrict(expectedExitCodeRaw);
  if (expectedExitCode === null) {
    return {
      feature,
      sourceFile,
      timestamp,
      command,
      exitCode: null,
      normalizedResult: "unparseable",
      expectedExitCode: 0,
    };
  }

  const normalizedResult: NormalizedResult = normalizeResult(
    exitCode,
    expectedExitCode,
  );
  return {
    feature,
    sourceFile,
    timestamp,
    command,
    exitCode,
    normalizedResult,
    expectedExitCode,
  };
}

/**
 * Read and parse one canonical evidence file.
 *
 * Mirrors Python `parse_verification_evidence_file`: reads `root/relativePath`
 * through the injected filesystem and parses it; a read failure propagates as a
 * thrown error (the collector catches it, matching the Python `OSError`
 * contract).
 *
 * @param params Filesystem, repo root, feature id, and repo-relative path.
 * @returns The parsed normalized evidence record.
 * @throws Error When the evidence file cannot be read.
 */
export function parseVerificationEvidenceFile(params: {
  fs: FileSystem;
  root: string;
  feature: string;
  relativePath: string;
}): VerificationEvidenceRecord {
  const { fs, root, feature, relativePath } = params;
  const normalizedRoot = toPosixPath(root).replace(/\/+$/u, "");
  const relPosix = toPosixPath(relativePath).replace(/^\/+/u, "");
  const markdown = fs.readTextFile(`${normalizedRoot}/${relPosix}`);
  return parseVerificationEvidenceMarkdown({
    feature,
    sourceFile: relPosix,
    markdown,
  });
}

/**
 * Parse an integer strictly, mirroring Python `int(str)` semantics for the
 * decimal strings evidence files contain.
 *
 * Python `int("0")` succeeds; `int("ok")` raises `ValueError`. This accepts an
 * optionally-signed run of decimal digits and rejects anything else.
 *
 * @param value Trimmed candidate string.
 * @returns The parsed integer, or `null` when not a valid integer.
 */
function parseIntegerStrict(value: string): number | null {
  if (!/^[+-]?\d+$/u.test(value)) {
    return null;
  }
  return Number.parseInt(value, 10);
}

/**
 * Convert an absolute POSIX path under `root` to a repo-relative POSIX path.
 *
 * @param root Normalized repository root (no trailing slash).
 * @param absolute Absolute POSIX path under `root`.
 * @returns The path relative to `root`, without a leading slash.
 */
function relativeToPosix(root: string, absolute: string): string {
  const normalized = toPosixPath(absolute);
  if (normalized.startsWith(`${root}/`)) {
    return normalized.slice(root.length + 1);
  }
  return normalized.replace(/^\/+/u, "");
}

/** Compare two strings by Unicode code point (Python `sorted` semantics). */
function compareCodePoint(left: string, right: string): number {
  if (left < right) {
    return -1;
  }
  if (left > right) {
    return 1;
  }
  return 0;
}

/**
 * Split text into lines the way Python `str.splitlines()` does.
 *
 * @param value Text to split.
 * @returns Lines without terminators.
 */
function splitLines(value: string): string[] {
  if (value === "") {
    return [];
  }
  const lines = value.split(/\r\n|\r|\n/u);
  if (
    lines.length > 0 &&
    lines[lines.length - 1] === "" &&
    /(\r\n|\r|\n)$/u.test(value)
  ) {
    lines.pop();
  }
  return lines;
}
